### PR TITLE
ARROW-5163 : [Gandiva] Cast timestamp/date are incorrectly evaluating year 0097 to 1997

### PR DESCRIPTION
**Pull request by shyambits2004**
Consider the length of the year string before modifying it. year string "0097" should be considered as 0097 and year string "97" should be considered as 1997.

### CODE REVIEW

File: cpp/src/gandiva/precompiled/time.cc [hunk 1]
```diff
@@ -634,7 +642,7 @@ timestamp castTIMESTAMP_utf8(int64_t context, const char* input, int32 length) {
   }
 
   // adjust the year
-  if (ts_fields[TimeFields::kYear] < 100) {
+  if (ts_fields[TimeFields::kYear] < 100 && year_str_len < 4) {
```
Comment (pitrou) [inline cpp/src/gandiva/precompiled/time.cc]: Shouldn't one set an error if year length is neither 2 nor 4?

**Review by pitrou — COMMENTED** [review]
LGTM, just a single question.

**Comment by pravindra** [comment]
sorry for the delay. Lgtm,  please add tests for the other cases where year is - single-digit, three-digit with leading zero, three-digit with no leading zero, five-digit ..

File: cpp/src/gandiva/precompiled/time.cc (hunk 1 above)
Comment (shyambits2004) [reply cpp/src/gandiva/precompiled/time.cc]: year length can vary from 1 - 4. Added test cases for that too.

**Review by pravindra — APPROVED** [review]

**Pull request closed without merging**